include/cudaq/Support/Device.h

```diff
@@ -0,0 +1,173 @@
+    Device device;
+    Qubit center = device.topology.createNode();
+    for (unsigned i = 1u; i < numQubits; ++i) {
+      device.topology.createNode();
+      device.topology.addEdge(center, Qubit(i));
+    }
+    return device;
+  }
+
+  /// Create a device with a grid topology.
+  ///
+  ///  0 -- 1 -- 2
+  ///  |    |    |
+  ///  3 -- 4 -- 5
+  ///  |    |    |
+  ///  6 -- 7 -- 8
+  ///
+  static Device grid(unsigned width, unsigned height) {
+    Device device;
+    for (unsigned i = 0u, end = width * height; i < end; ++i)
+      device.topology.createNode();
+    for (unsigned x = 0u; x < width; ++x) {
+      for (unsigned y = 0u; y < height; ++y) {
+        unsigned base = x + (y * width);
+        Qubit q0(base);
+        if (x < width - 1)
+          device.topology.addEdge(q0, Qubit(base + 1));
+        if (y < height - 1)
+          device.topology.addEdge(q0, Qubit(base + width));
+      }
+    }
+    return device;
+  }
+
+  /// TODO: Implement a method to load device info from a file.
+
+  /// Returns the number of physical qubits in the device.
+  unsigned getNumQubits() const { return topology.getNumNodes(); }
+
+  /// Returns the distance between two qubits.
+  unsigned getDistance(Qubit src, Qubit dst) const {
+    unsigned pairID = getPairID(src.index, dst.index);
+    return src == dst ? 0 : shortestPaths[pairID].size() - 1;
+  }
+
+  mlir::ArrayRef<Qubit> getNeighbours(Qubit src) const {
+    return topology.getNeighbours(src);
+  }
+
+  bool areConnected(Qubit q0, Qubit q1) const {
+    return getDistance(q0, q1) == 1 ? true : false;
+  }
+
+  /// Returns a shortest path between two qubits.
+  Path getShortestPath(Qubit src, Qubit dst) const {
+    unsigned pairID = getPairID(src.index, dst.index);
+    if (src.index > dst.index)
+      return Path(llvm::reverse(shortestPaths[pairID]));
+    return Path(shortestPaths[pairID]);
+  }
+
+  void dump(llvm::raw_ostream &os = llvm::errs()) const {
+    os << "Graph:\n";
+    topology.dump(os);
+    os << "\nShortest Paths:\n";
+    for (unsigned src = 0; src < getNumQubits(); ++src)
+      for (unsigned dst = 0; dst < getNumQubits(); ++dst) {
+        auto path = getShortestPath(Qubit(src), Qubit(dst));
+        os << '(' << src << ", " << dst << ") : {";
+        llvm::interleaveComma(path, os);
+        os << "}\n";
+      }
+  }
+
+private:
+  using PathRef = mlir::ArrayRef<Qubit>;
+
+  unsigned getPairID(unsigned u, unsigned v) const {
+    if (u > v)
+      std::swap(u, v);
+    return (u * getNumQubits()) - (((u - 1) * u) / 2) + v - u;
+  }
+
+  void computeAllPairShortestPaths() {
+    std::size_t numNodes = topology.getNumNodes();
+    shortestPaths.resize(numNodes * (numNodes + 1) / 2);
+    mlir::SmallVector<Qubit> path;
+    for (unsigned n = 0; n < numNodes; ++n) {
+      auto parents = getShortestPathsBFS(topology, Qubit(n));
+      // Reconstruct the paths
+      for (auto m = n + 1; m < numNodes; ++m) {
+        path.clear();
```

Comment:
nit: seems like you can move `path` here in the body of the loop and give it a worst case size based on the topology and it would be more efficient. (Space would be preallocated; clearing automatic.)

Reply:
Moving it inside the loop would make it allocate memory each time through the loop, no? In any case, yes, I think we can initialize the vector's size according to the worst-case length (`numNodes`).

Reply:
Worst-case allocation set in 9f60585e.